src/AWS/Lambda/Events/Kafka.hs

```diff
@@ -0,0 +1,249 @@
+    "SelfManagedKafka" -> pure SelfManagedKafka
+    t -> fail $ "unrecognised EventSource: \"" <> show t <> "\""
+
+instance ToJSON EventSource where
+  toJSON = \case
+    AwsKafka -> String "aws:kafka"
+    SelfManagedKafka -> String "SelfManagedKafka"
+
+-- | Convenience alias: most of the time you will parse the records
+-- straight into some app-specific structure.
+type Record = Record' ByteString
+
+-- | Records from a Kafka event. This is 'Traversable', which means
+-- you can do things like parse a JSON-encoded payload:
+--
+-- @
+-- 'traverse' 'decodeStrict' :: 'FromJSON' a => Record ByteString -> Maybe (Record a)
+-- @
+data Record' a = Record {
+  topic :: !Text,
+  partition :: !Int32,
+  offset :: !Int64,
+  timestamp :: !Timestamp,
+  -- | NOTE: there can be multiple headers for a given key.
+  headers :: !(Vector Header),
+  key :: !(Maybe ByteString),
+  value :: !(Maybe a)
+} deriving (Eq, Show, Generic, Functor, Foldable, Traversable)
+
+-- | Decodes base64-encoded keys and values, where present.
+instance a ~ ByteString => FromJSON (Record' a) where
+  parseJSON = withObject "Record" $ \o -> do
+    topic <- o .: "topic"
+    partition <- o .: "partition"
+    offset <- o .: "offset"
+    timestamp <- parseTimestamp o
+    headers <- o .: "headers"
+    key <- fmap (B64.decodeLenient . TE.encodeUtf8) <$> o .:? "key"
+    value <- fmap (B64.decodeLenient . TE.encodeUtf8) <$> o .:? "value"
+    pure Record { topic, partition, offset, timestamp, headers, key, value }
+
+-- | Encodes keys and values into base64.
+instance a ~ ByteString => ToJSON (Record' a) where
+  toJSON r = object $ concat
+    [
+      [ "offset" .= offset r
+      , "partition" .= partition r
+      , "topic" .= topic r
+      , "headers" .= headers r
+      ]
+    , unparseTimestamp (timestamp r)
+    , catMaybes
+      [ ("key" .=) . TE.decodeUtf8 . B64.encode <$> key r
+      , ("value" .=) . TE.decodeUtf8 . B64.encode <$> value r
+      ]
+    ]
+
+-- | AWS serialises record headers to JSON as an array of
+-- objects. From their docs:
+--
+-- @
+-- "headers":[{"headerKey":[104, 101, 97, 100, 101, 114, 86, 97, 108, 117, 101]}]
+-- @
+--
+-- Note:
+--
+-- >>> map chr [104, 101, 97, 100, 101, 114, 86, 97, 108, 117, 101]
+-- "headerValue"
+data Header = Header !Text !ByteString deriving (Eq, Show, Generic)
+
+instance FromJSON Header where
+  parseJSON = withObject "header" $ \o ->
+    case KM.toList o of
+      [(key, value)] -> Header (keyToText key) . B.pack <$> parseJSON value
+      [] -> fail "Unexpected empty object"
+      _ -> fail "Unexpected additional keys in object"
+
+instance ToJSON Header where
+  toJSON (Header key value) = object [keyFromText key .= B.unpack value]
+
+-- | Kafka timestamp types, derived from the Java client's enum at:
+-- https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/record/TimestampType.java
+data Timestamp = NoTimestampType | CreateTime !UTCTime | LogAppendTime !UTCTime
```

Comment:
Often I find a tension between records that represent serialisation and records that are used as domain objects, and because they start off so similarly you get a type that tries to do both things badly (e.g., too much automagic conversion). To avoid this, I find it useful to be deliberate about what role a record type is expected to play. In `hal`, I feel that these records are primarily about serialisation, so I've tried to follow the data format closely and allowed library users parse into some other structure.

Your suggestion is more idiomatic Haskell but doesn't have the obvious correlation between enum names in the Kafka codebase and constructor names in `Timestamp`.

Reply:
(Apologies, this was a half finished comment for far too long)

Ah, I see that now--I originally thought this was trying to replicate the Java library based on the link to it, but see that it's more that the Java library is simply based on the serialization (or vise versa).  This is definitely an area where drawing the line is tough to do.  I think other places in `hal` have leaned heavier into idiomatic decoding, but the benefit of the clear connection to other docs seems like a reasonable benefit, and this is actually pretty close to idiomatic here.  I think you're right that this is the way to go.